### PR TITLE
New semantic analyzer: add tests for casts and type application

### LINKED ISSUE #6291: New semantic analyzer: type application
Support type applications such as in `x = C[int]()`.

### PR DESCRIPTION
**Pull request by msullivan**
It seems like they already work fine?

Closes #6291, #6292.

### CODE REVIEW

File: test-data/unit/check-newsemanal.test [hunk 1]
```diff
@@ -1856,3 +1858,47 @@ reveal_type(y[0])  # E: Revealed type is 'builtins.int'
...
+
+x = cast('C', None)
+class A:
+    def foo(self) -> None:
+        self.x = cast('C', None)
+
+reveal_type(x)  # E: Revealed type is '__main__.C'
+reveal_type(A().x)  # E: Revealed type is '__main__.C'
+
+class C(A): ...
```
Comment (ilevkivskyi) [inline test-data/unit/check-newsemanal.test]: Maybe add couple of tests that use special forms (type aliases, named tuples, etc)?

File: test-data/unit/check-newsemanal.test [hunk 2]
```diff
@@ -1856,3 +1858,47 @@ reveal_type(y[0])  # E: Revealed type is 'builtins.int'
...
+
+x = C['A']()
+reveal_type(x)  # E: Revealed type is '__main__.C[__main__.A*]'
+
+class A: ...
+
+[case testNewAnalyzerApplicationForward3]
+from typing import Generic, TypeVar
+
+x = C[A]()
```
Comment (ilevkivskyi) [inline test-data/unit/check-newsemanal.test]: Maybe add a test where the upper bound (also forward one) is violated?

File: test-data/unit/check-newsemanal.test [hunk 3]
```diff
@@ -1476,6 +1476,8 @@ strict_optional = True
 [[mypy-b]
 strict_optional = False
 
+-- ]]
```
Comment (ilevkivskyi) [inline test-data/unit/check-newsemanal.test]: Btw, why is this needed? We a bunch of unbalanced `[` in other test files.

File: test-data/unit/check-newsemanal.test [hunk 4]
```diff
@@ -1856,3 +1858,78 @@ reveal_type(y[0])  # E: Revealed type is 'builtins.int'
...
+reveal_type(x)  # E: Revealed type is 'builtins.int'
+
+C = int
+
+[case testNewAnalyzerCastForward2]
+from typing import cast, NamedTuple
+
+x = cast('C', None)
+
+reveal_type(x)  # E: Revealed type is '__main__.C'
```
Comment (ilevkivskyi) [inline test-data/unit/check-newsemanal.test]: This should be a tuple, the test failure is likely because of this.

File: test-data/unit/check-newsemanal.test (hunk 3 above)
Comment (msullivan) [reply test-data/unit/check-newsemanal.test]: Yeah, and I usually fix them when I run into them :P 
They mess up editor indentation, which I find super annoying
